src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesFolderProperty.java

```diff
@@ -0,0 +1,284 @@
+            p.setInherited(false);
+
+            ps.add(p);
+        }
+
+        return ps;
+    }
+
+    private static String PREFIX_USAGE_PERMISSION = "usage-permission-";
+
+    @Override
+    public AbstractFolderProperty<?> reconfigure(StaplerRequest req, JSONObject form) throws FormException {
+        if (form == null) {
+            return null;
+        }
+
+        // ignore modifications silently and return the unmodified object if the user
+        // does not have the ADMINISTER Permission
+        if (!userHasAdministerPermission()) {
+            return this;
+        }
+
+        try {
+            KubernetesFolderProperty newKubernetesFolderProperty = new KubernetesFolderProperty();
+
+            Set<String> inheritedGrants = new HashSet<>();
+            collectAllowedClouds(inheritedGrants, getOwner().getParent());
+
+            JSONArray names = form.names();
+            if (names != null) {
+                for (Object name : names) {
+                    String strName = (String) name;
+
+                    if (strName.startsWith(PREFIX_USAGE_PERMISSION) && form.getBoolean(strName)) {
+                        String cloud = StringUtils.replaceOnce(strName, PREFIX_USAGE_PERMISSION, "");
+
+                        if (isUsageRestrictedKubernetesCloud(Jenkins.getInstance().getCloud(cloud))
+                                && !inheritedGrants.contains(cloud)) {
+                            newKubernetesFolderProperty.getPermittedClouds().add(cloud);
+                        }
+                    }
+                }
+            }
+
+            return newKubernetesFolderProperty;
+        } catch (JSONException jsonException) {
+            LOGGER.severe(String.format("reconfigure failed: %s", jsonException.getMessage()));
+            jsonException.printStackTrace();
+
+            return this;
+        }
+    }
+
+    /**
+     * Recursively collect all allowed clouds from this folder and its parents.
+     * 
+     * @param allowedClouds
+     *            This Set contains all allowed clouds after returning.
+     * @param itemGroup
+     *            The itemGroup to inspect.
+     */
+    public static void collectAllowedClouds(Set<String> allowedClouds, ItemGroup<?> itemGroup) {
+        if (itemGroup instanceof AbstractFolder) {
+            AbstractFolder<?> folder = (AbstractFolder<?>) itemGroup;
+            KubernetesFolderProperty kubernetesFolderProperty = folder.getProperties()
+                    .get(KubernetesFolderProperty.class);
+
+            if (kubernetesFolderProperty != null) {
+                allowedClouds.addAll(kubernetesFolderProperty.getPermittedClouds());
+            }
+
+            collectAllowedClouds(allowedClouds, folder.getParent());
+        }
+    }
+
+    private static List<KubernetesCloud> getUsageRestrictedKubernetesClouds() {
+        List<KubernetesCloud> k8sClouds = new ArrayList<>();
+
+        for (Cloud cloud : Jenkins.getInstance().clouds) {
```

Comment:
Could use `Jenkins.getInstance().clouds.getAll(KubernetesCloud.class)` then no need to cast.